### ISSUE TITLE
Logging inside of Docker doesn't work

### ISSUE THREAD

**Issue by Firehed** [spiral/roadrunner#21]
In a comment in #12, it came up that anything sent to STDERR should get relayed to Docker and displayed, but in my experience that's simply not the case. I've tried the following things, none of which work:

- `ini_set('display_errors', 'stderr');`
- `fwrite(STDOUT, 'some message');` (this can crash the relay process, since it corrupts the packet format)
- `fwrite(STDERR, 'some message');`

File-based logging works fine, but doesn't actually help without a tremendous amount of complexity (Docker is, generally speaking, watching the top-level process's `STDOUT` and `STDERR`).

The only logs I've been able to get (by using `rr serve -v -d`) are the access logs, and they get goofy formatting due to attempted colorization.

It would be immensely preferable if this worked the same way PHP-FPM does; i.e. STDOUT and STDERR "just work" (a random var_dump/exit debugger in a script won't disappear, etc), but I can't think of a good way to support this while keeping supports for sockets working. The pipes connection method could be adapted to make STDERR work as expected, but it wouldn't solve the STDOUT thing without quite a lot of work (I have a couple of napkin-quality ideas about this, but don't know Go too well)

**Comment by wolfy-j** [comment]
OK, actually bypassing STDERR in realtime appeared to be super easy, working on this improvement. I'm not going to do anything with STDOUT, for now, to keep worker relays consistent.

**Comment by wolfy-j** [comment]
@Firehed check 1.0.3, RR now exposes stderr. I will check the formatting issue for Docker in next version, debug mode was not purposed for that.

**Comment by Firehed** [comment]
Will do - thanks for looking into this!

I've been spending some time thinking about a way to "fully" solve this, but I believe part of it is inherent to having a long-running process as compared to e.g. php-fpm's run-once model (it's likely that e.g. PHP-PM is similarly impacted).

It's probably possible to get very close using some bizarre IPC voodoo (e.g. have a fourth FD open that accomplishes what [this](https://github.com/spiral/goridge/blob/master/php-src/StreamRelay.php#L60) does), but that may not be worth the complexity and it might still not be reliable or even fully solve the problem.

To answer your previous question:

> Are you ok using unix sockets in the docker? 

Yes - but an example in the README would be beneficial here. I haven't yet attempted to do so, but it seems like you'd need to configure the connection dynamically per-worker otherwise all workers would just fight over a single path/port. Or is that not the case?

**Comment by wolfy-j** [comment]
> It's probably possible to get very close using some bizarre IPC voodoo (e.g. have a fourth FD open that accomplishes what this does), but that may not be worth the complexity and it might still not be reliable or even fully solve the problem.

I've been thinking about the same today, opening an additional channel of communication might help. I will think about using STDERR in future for that (optionally).

However, there is much-much easier way to communicate with parent service - use embedded RPC service. We are using it at production to forward logs and stats to Prometheus directly from PHP script.

> Yes - but an example in the README would be beneficial here. I haven't yet attempted to do so, but it seems like you'd need to configure the connection dynamically per-worker otherwise all workers would just fight over a single path/port. Or is that not the case?

No, RR supports proper multiplexing over TCP and Unix.

The change is pretty simply, set relay as "tcp://:222" or "unix://socket.unix" and do the same on worker end.

**Comment by wolfy-j** [comment]
@Firehed can you give me a screenshot of `they get goofy formatting due to attempted colorization`?

**Comment by wolfy-j** [comment]
```yaml
relay:    "unix://socket.unix"
relayTimeout: 30000000000
```

RelayTimeout is required for socket based connections to workers as allocation is not immediate compared to pipes.

**Comment by Firehed** [comment]
>  I will think about using STDERR in future for that (optionally).

I'd strongly suggest not abusing STDERR for that - what I'm thinking would be in addition to the expected use of STDIN/OUT/ERR.

_But_ I hadn't considered the use of the embedded RPC service - that seems like a much better solution. Granted it's probably almost the same actual thing under the hood, but feels way less janky.

I'll send a screenshot of the formatting shortly - need to get back on the correct system. It's either incomplete or incompatible color codes, and could be specific to using `docker-compose` which I was testing with. If I run the image directly (`docker run --rm -it my-roadrunner-tag`) it's fine.

**Comment by wolfy-j** [comment]
RPC works over sockets and does not have same limitations as pipes, it is also much faster than http layer (no extra PSR at top), so you can reach 100K-200K rps easily.

**Comment by Firehed** [comment]
Now that I'm back at my work machine, here's the formatting issue (after updating to 1.0.3, which did indeed solve the logging thing):

```
rr_1            | time="2018-06-25T16:40:53Z" level=debug msg="[rpc]: started"
rr_1            | time="2018-06-25T16:40:53Z" level=debug msg="[http]: started"
rr_1            | time="2018-06-25T16:40:53Z" level=debug msg="[static]: started"
rr_1            | time="2018-06-25T16:42:52Z" level=info msg="\x1b[32m200\x1b[0m \x1b[1;97mGET\x1b[0m http://localhost:9999/meta/all"
rr_1            | time="2018-06-25T16:43:17Z" level=info msg="\x1b[31m500\x1b[0m \x1b[1;97mGET\x1b[0m http://localhost:9999/meta/al"
rr_1            | time="2018-06-25T16:43:17Z" level=warning msg="[2018-06-25T16:43:17+00:00] [error] OutOfBoundsException: Endpoint not found in /www/vendor/firehed/api/src/Dispatcher.php:211\nStack trace:\n#0 /www/vendor/firehed/api/src/Dispatcher.php(129): Firehed\\API\\Dispatcher->getEndpoint()\n#1 /www/rr-entrypoint.php(36): Firehed\\API\\Dispatcher->dispatch()\n#2 {main}"
```

The two items of note are the color codes in the http code info messages, and newlines being printed as a string literal in the would-be stack trace (which is from a PSR logger writing to STDERR). This is while running `rr serve -v -d`. It would be nice if the child logging was a bit more configurable from the rr config file, but the immediate issue of not being able to get anything out of workers seems resolved.

**Comment by wolfy-j** [comment]
Yeah, I see that too, I will provide an option to force logrus to output in more convenient format.

**Comment by wolfy-j** [comment]
Fixed in v1.0.5
